llvm/lib/Transforms/Scalar/Scalarizer.cpp

```diff
@@ -689,8 +698,10 @@ bool ScalarizerVisitor::splitBinary(Instruction &I, const Splitter &Split) {
   return true;
 }
 
-static bool isTriviallyScalariable(Intrinsic::ID ID) {
-  return isTriviallyVectorizable(ID);
+bool ScalarizerVisitor::isTriviallyScalarizable(Intrinsic::ID ID) {
+
+  return TTI->isTargetIntrinsicTriviallyScalarizable(ID) ||
+         isTriviallyVectorizable(ID);
```

Comment:
Swap these.

Reply:
Also, should this check isTargetIntrinsic() first?

Reply:
When I was first designing this feature I had 
```c++
 bool isTargetIntrinsicTriviallyScalarizable(Intrinsic::ID ID) {
  if (!Function::isTargetIntrinsic(ID))
    return false;
...
}
```
Hence why the names are so similar. However, after I implemented the DirectX part I realized that i'd be returning false anyways if it wasn't a dx intrinsic.

What we could do since I switched the order is to add an assert?

Reply:
I think this check should be here in Scalarizer, so that isTargetIntrinsicTriviallyScalarizable() doesn't get called with a non-target intrinsic -- it shouldn't be able to influence whether generic intrinsics are scalarizable. So something like

```
if (isTriviallyVectorizable(ID))
  return true;
return Function::isTargetIntrinsic(ID) && TTI->isTargetIntrinsicTriviallyScalarizable(ID);
```
is what I'd do.